### DISCUSSION TITLE
How does the watchdog work?

### DISCUSSION THREAD

**Discussion by nicklasb** [feilipu/Arduino_FreeRTOS_Library#117] [Q&A] [answered]
Hi!
I am a bit confused by how the watchdog works in Arduino FreeRTOS.
As I have to run my Unity tests in a task to use delay and stuff like that, it seems that I am triggering the watchdog. 
If I try to set the watchdog, changing portUSE_WDTO instead seems to cause other things to change that also may cause the watchdog to trigger. 

TBH, it is very difficult to troubleshoot this, at least to me, so if anyone have an idea what to do, I'd be very happy. :-)

**Comment by feilipu** [comment, marked as answer]
I'm not 100% sure how to answer your question, or perhaps to address your underlying concern.

The AVR ATmega Watchdog Timer (WDT) is a simple counter that can be configured on roll-over to a) reset the MCU, b) trigger an interrupt, c) trigger and interrupt and if the interrupt is unserviced reset the MCU on the next counter roll-over. The counter is incremented by an integrated oscillator, independent of any other oscillators in the MCU, which runs at nominally 128kHz.

In Arduino FreeRTOS the WDT is configured to firstly interrupt at roll-over and then at the second roll-over then reset the MCU. If you keep the global interrupt disabled for extended times then you will find that your MCU is resetting, but really 15ms + 15ms at 16MHz is a very long time.

There are [issue](https://github.com/feilipu/Arduino_FreeRTOS_Library/issues/113) and [issue ](https://github.com/feilipu/Arduino_FreeRTOS_Library/issues/114)here on the Arduino delay function, and the FreeRTOS delay function.

Perhaps they will cast some light on the question for you too?

**Comment by nicklasb** [comment]
@feilipu 
No, your hit the head on the nail, I realize that the question wasn't very clear.
Ok, it was as i feared then. 
I do agree that 15 ms is a fairly long time usually.. but it isn't always. I am simulating communications and often need logging (TDD becomes hard if not) and they can often take longer with slower media. Also some libraries aren't always calling vTaskDelay very often. Actually, completely disabling it would be helpful for some test cases.

Either way I will have to see if I can work around it, somehow.  Thanks for your help!

(Background is that I try to run the same code on ESP32, Linux and Arduino aswell...just because that is even messier. :-) )
